src/consensus/mod.rs

```diff
@@ -47,60 +55,106 @@ impl<'a, C, B> FastPaxos<'a, C, B> {
         let mut paxos_delay = Delay::new(Instant::now() + self.get_random_delay()).fuse();
 
-        // a hacky way to schedule a task after the delay
         async {
             paxos_delay.await;
             self.start_classic_round().await;
-        }.await;
+        }
+            .await;
+
+        let (tx, rx) = oneshot::channel();
+
+        let kind = proto::RequestKind::Consensus(proto::Consensus::FastRoundPhase2bMessage(
+            proto::FastRoundPhase2bMessage {
+                sender: self.my_addr.clone(),
+                config_id: self.config_id,
+                endpoints: proposal,
+            },
+        ));
+
+        let request = Request::new(tx, kind);
 
-        self.broadcast.broadcast(Phase2bMessage {}).await;
+        self.broadcast.broadcast(request).await;
+        rx.await;
 
+        // TODO: better error type
         Ok(())
     }
 
-    pub async fn handle_message(&self, request: Request) -> Result<Response> {
+    pub async fn handle_message(&mut self, request: Request) -> Result<Vec<Endpoint>> {
         match request.kind() {
             proto::RequestKind::Consensus(req_type) => {
-                if let Some(paxos) = self.paxos {
+                if self.paxos.is_some() {
                     match req_type {
-                        proto::Consensus::FastRoundPhase1bMessage(_) => self.handle_fast_round(request).await,
-                        proto::Consensus::Phase1aMessage(_) => paxos.handle_phase_1a(request).await,
-                        proto::Consensus::Phase1bMessage(_) => paxos.handle_phase_1b(request).await,
-                        proto::Consensus::Phase2aMessage(_) => paxos.handle_phase_2a(request).await,
-                        proto::Consensus::Phase2bMessage(_) => paxos.handle_phase_2b(request).await,
+                        proto::Consensus::FastRoundPhase2bMessage(msg) => {
+                            self.handle_fast_round(msg).await
+                        }
+                        // TODO: need to figure out communication b/w paxos and fast paxos
+                        _ => unimplemented!(),
                     }
                 } else {
                     Err(Error::new_unexpected_request(None))
                 }
-            },
+            }
             _ => Err(Error::new_unexpected_request(None)),
         }
     }
 
-    async fn handle_fast_round(&self, request: Request) -> crate::Result<Response> {
-        unimplemented!()
+    async fn handle_fast_round<'b>(
+        &'b mut self,
+        request: &'b proto::FastRoundPhase2bMessage,
+    ) -> crate::Result<Vec<Endpoint>> {
+        if request.config_id != self.config_id {
+            return Err(Error::new_unexpected_request(None));
+        }
+
+        if self.votes_received.contains(&request.sender) {
+            return Err(Error::new_unexpected_request(None));
+        }
+
+        if (self.decided.load(Ordering::SeqCst)) {
```

Comment:
same here I think we can make this a regular `bool`.

Reply:
I don' know if that's a good idea, as there could be different futures reading from the `decided` variable.